compiler/rustc_mir/src/const_eval/eval_queries.rs

```diff
@@ -248,18 +263,29 @@ pub fn eval_to_allocation_raw_provider<'tcx>(
     // computed. For a large percentage of constants that will already have succeeded. Only
     // associated constants of generic functions will fail due to not enough monomorphization
     // information being available.
-
-    // In case we fail in the `UserFacing` variant, we just do the real computation.
-    if key.param_env.reveal() == Reveal::All {
-        let mut key = key;
-        key.param_env = key.param_env.with_user_facing();
-        match tcx.eval_to_allocation_raw(key) {
-            // try again with reveal all as requested
-            Err(ErrorHandled::TooGeneric) => {}
-            // deduplicate calls
-            other => return other,
+    match key.param_env.reveal() {
+        Reveal::Selection => {}
+        Reveal::UserFacing => {
+            let mut key = key;
+            key.param_env = key.param_env.with_reveal_selection();
+            match tcx.eval_to_allocation_raw(key) {
+                Err(ErrorHandled::Silent) => {}
```

Comment:
If we stick the actual error into the `Silent` variant, then we don't have to re-run. Not sure if this rerunning has an impact on perf

Reply:
wasn't the concern by you and @RalfJung here that doing so requires us to pass these errors through query boundaries which is somewhat difficult? :thinking: 

I didn't look into that possibility myself yet

Reply:
Oh I misread... I thought this was for already-reported, too-generic and unknown-layout. but it's the other way around.

`Reveal::UserFacing` won't ever error on these, so this forwarding is correct here... I still wonder how to reduce the number of query calls... Maybe let's benchmark before trying to optimize here though